File: kotlin/sdk/src/commonMain/kotlin/com/basecamp/sdk/services/CampfireIndex.kt
package com.basecamp.sdk.services

import com.basecamp.sdk.AccountClient
import com.basecamp.sdk.BasecampException
import com.basecamp.sdk.PaginationOptions
import com.basecamp.sdk.generated.campfires
import com.basecamp.sdk.generated.projects
import kotlinx.coroutines.CompletableDeferred
import kotlinx.coroutines.NonCancellable
import kotlinx.coroutines.currentCoroutineContext
import kotlinx.coroutines.ensureActive
import kotlinx.coroutines.isActive
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.coroutines.withContext
import kotlin.coroutines.cancellation.CancellationException
import kotlin.time.TimeSource

/**
 * How long a cached discovery source — a bucket's project dock, the account's
 * Campfire listing — is reused before it is read again.
 */
const val CAMPFIRE_INDEX_TTL_MILLIS: Long = 10 * 60 * 1000

/**
 * Bounds the refresh-on-miss: a line found under no candidate re-reads the
 * cached sources, but not more often than this per source, so a run of
 * unresolvable lines cannot turn into a listing per line.
 * [BasecampException.RecordingSummaryFailure.refreshed] says whether the floor
 * applied.
 */
internal const val CAMPFIRE_INDEX_MIN_REFRESH_MILLIS: Long = 30 * 1000

/**
 * Bounds how many Campfires one `summarize` call tries, across both sources and
 * the refresh. A project has one Campfire and a handful of pings; a bucket past
 * this bound is not a shape BC3 produces, and the call reports
 * `campfire_discovery_incomplete` rather than calling the rest absent.
 */
const val MAX_CAMPFIRE_CANDIDATES: Int = 50

/**
 * Caps the account-wide Campfire listing the fallback source reads. A listing
 * that overflows it is not cached and the call reports
 * `campfire_discovery_incomplete`: the dock covers every project, so the listing
 * only ever serves the leftover, and an account with more Campfires than this
 * should not pay a full walk per ten minutes for it.
 */
const val MAX_CAMPFIRE_LISTING: Int = 1000

/**
 * Bounds each discovery cache's entry count. The dock cache holds one snapshot
 * per bucket consulted within the TTL: a connector listening across every
 * project an agent can see touches hundreds of buckets, not thousands, and a
 * snapshot is a handful of ids, so 1024 is generous headroom at a few hundred
 * KB, and the bound exists so that a process alive for weeks can never grow past
 * it whatever it sees. When the bound is reached the oldest-fetched entries go
 * first, deterministically.
 */
private const val CAMPFIRE_INDEX_MAX_ITEMS = 1024

/**
 * The cache's clock: milliseconds since an arbitrary fixed origin, read from a
 * MONOTONIC source rather than the wall clock.
 *
 * Go gets this for free — a `time.Time` from `time.Now()` carries a monotonic
 * reading and `Sub` uses it — and the cache depends on it in three places that
 * all fail quietly under a backwards wall-clock step: an entry whose computed
 * age goes negative outlives its TTL and survives the sweep, a refresh is
 * declined below a floor it has actually passed, and the fetched-time comparison
 * that decides `refreshed` (and therefore the stale-candidate report) reads the
 * wrong way. None of those raise; they just answer wrongly.
 */
internal fun monotonicMillis(): Long = MONOTONIC_ORIGIN.elapsedNow().inWholeMilliseconds

private val MONOTONIC_ORIGIN = TimeSource.Monotonic.markNow()

/**
 * What a cache read hands back: the value, when it was fetched, and whether it
 * predated the call (as opposed to being loaded during it, by this caller or by
 * one it waited on). The fetch time is what lets a caller tell a snapshot it
 * already consulted from a newer one, whoever loaded it.
 */
internal class TtlHit<V>(val value: V, val fetched: Long, val cached: Boolean)

/**
 * A per-key cache with single-flight loading: concurrent callers for one key
 * wait on the one load in progress rather than loading again, a failed load
 * leaves the previous value in place, and a refresh is honoured only once the
 * value is older than a floor.
 */
internal class TtlCache<K, V>(
    private val now: () -> Long,
    private val ttlMillis: Long,
    private val floorMillis: Long,
    private val maxItems: Int,
) {
    private class Entry<V>(val value: V, val fetched: Long, val seq: Long)

    private class Load<V> {
        val done = CompletableDeferred<V>()

        /** Set at publication, so the loader's own hit reports a real fetch time. */
        var fetched: Long = 0

        /**
         * Records that the failure is attributable to the loading caller's own
         * cancellation: its job was no longer active when the load ended AND the
         * failure is a cancellation.
         *
         * Both halves are needed, and the second one is why: Ktor's
         * `HttpRequestTimeoutException` IS a `CancellationException` and arrives
         * with the owner's job still active. Classifying on the type alone would
         * make that the owner's cancellation and turn one timed-out request into
         * one request per waiter.
         *
         * It is a PROXY for Go's `callerDone`, not a translation of it. Go asks
         * whether the error is the context's own (`errors.Is(err, ctxErr)`);
         * this asks whether the job is still active, which answers the same
         * question everywhere except one race — a transport timeout landing in
         * the same instant as an unrelated cancellation of the owner reads as
         * owner-attributed here and as shared in Go. The cost is bounded to one
         * extra load by the `reacquired` latch, and Go's own comment concedes an
         * overlap it cannot separate either.
         */
        var ownerCancelled: Boolean = false
    }

    private val mutex = Mutex()
    private val entries = mutableMapOf<K, Entry<V>>()
    private val inflight = mutableMapOf<K, Load<V>>()

    /** Publications so far; stamps entries so eviction order is total. */
    private var seq: Long = 0

    /**
     * Returns the value for [key], loading it when absent or older than the TTL
     * — or, with [refresh] set, older than the floor.
     */
    suspend fun get(key: K, refresh: Boolean, load: suspend () -> V): TtlHit<V> {
        var reacquired = false
        while (true) {
            // A cancelled caller gets no answer, cached or otherwise: the mutex
            // below may be free, and a fresh hit would otherwise let a cancelled
            // call run to a verdict without ever suspending.
            currentCoroutineContext().ensureActive()
            var fresh: TtlHit<V>? = null
            var pending: Load<V>? = null
            var owned: Load<V>? = null
            mutex.withLock {
                val entry = entries[key]
                if (entry != null) {
                    val age = now() - entry.fetched
                    if (age < ttlMillis && (!refresh || age < floorMillis)) {
                        fresh = TtlHit(entry.value, entry.fetched, cached = true)
                    }
                }
                if (fresh == null) {
                    val existing = inflight[key]
                    if (existing != null) {
                        pending = existing
                    } else {
                        val load0 = Load<V>()
                        inflight[key] = load0
                        owned = load0
                    }
                }
            }
            fresh?.let { return it }

            val waited = pending
            if (waited != null) {
                val value = try {
                    waited.done.await()
                } catch (e: CancellationException) {
                    // This caller's OWN cancellation is its own to throw, and it
                    // takes precedence over anything the load did.
                    currentCoroutineContext().ensureActive()
                    // The load ran under the loading caller's coroutine. If that
                    // caller was cancelled, the failure is its, not this one's: a
                    // waiter whose own job is live goes round again and loads for
                    // itself (the key is free, so it becomes the loader and any
                    // other waiters queue behind it — one load, not a stampede).
                    // Once only, so a run of cancelled owners cannot become a
                    // queue of sequential loads behind one waiter.
                    if (waited.ownerCancelled && !reacquired) {
                        reacquired = true
                        continue
                    }
                    if (waited.ownerCancelled) {
                        // Past the one retry, and the failure belongs to a job
                        // that is not this one. Relaying a foreign
                        // CancellationException would end THIS live coroutine as
                        // "cancelled" and do it silently: JobSupport.childCancelled
                        // short-circuits on a CancellationException, so no
                        // CoroutineExceptionHandler ever runs and a `launch { }`
                        // caller loses the failure entirely. Go hands its waiter an
                        // ordinary error value; this is the nearest thing Kotlin
                        // has to that.
                        throw BasecampException.Api(
                            "the campfire discovery read this call was waiting on was cancelled by the caller that started it",
                            httpStatus = null,
                            hint = "retry; another caller's cancellation is not this one's failure",
                            cause = e,
                        )
                    }
                    // Anything else is the load's OWN failure and is shared —
                    // a transport timeout included, which Ktor spells as a
                    // CancellationException. The owner sees that same shape
                    // natively, so a waiter seeing it too is the honest answer.
                    throw e
                }
                // The load this call waited on is this call's load: hand its value
                // over as fresh, not as something that predated the call — read off
                // the load record, so a sweep or the bound evicting the entry in the
                // meantime cannot take it from this waiter.
                return TtlHit(value, waited.fetched, cached = false)
            }

            val load0 = requireNotNull(owned)
            // Every exit from here releases the key. `publishSuccess` is INSIDE
            // the try, so a failure in publication itself lands in the catch and
            // republishes as a failure rather than leaking the slot; and the
            // catch is on Throwable, so an Error and a cancellation are covered
            // too. That totality is what Go gets from its deferred publish, and
            // it is load-bearing rather than tidy: waiters wait with no timeout,
            // so a slot never released parks every later caller for that key for
            // the life of the client — and the listing's key is the bare account
            // id, which makes that an account rather than a bucket.
            //
            // A `finally` would add nothing reachable on top of that. The only
            // ways past a Throwable catch are a throw from inside the catch
            // itself, and a coroutine that is never resumed at all — and a
            // coroutine that is never resumed does not run its `finally` either,
            // so the guard could not help where it would be needed. Go's
            // goroutine has the identical hole.
            try {
                val value = load()
                publishSuccess(key, load0, value)
                return TtlHit(value, load0.fetched, cached = false)
            } catch (t: Throwable) {
                load0.ownerCancelled = t is CancellationException && !currentCoroutineContext().isActive
                publishFailure(key, load0, t)
                throw t
            }
        }
    }

    /**
     * Returns the cached value for [key] when one is within the TTL, without
     * loading. It lets a caller consult what a source already holds before
     * deciding whether to pay for a fetch of it.
     */
    suspend fun peek(key: K): TtlHit<V>? = mutex.withLock {
        val entry = entries[key] ?: return@withLock null
        if (now() - entry.fetched >= ttlMillis) return@withLock null
        TtlHit(entry.value, entry.fetched, cached = true)
    }

    /**
     * Publication is NonCancellable, on both paths and for the same reason Go
     * publishes from a `defer`: it always runs. A cancellation landing between
     * the loader returning and the key being released would otherwise throw out
     * of `withLock`, leaving the key in flight and its waiters suspended on a
     * deferred nobody will ever complete — a key poisoned for the life of the
     * client. The failure path needs it most, since it is reached AFTER a
     * cancellation.
     */
    private suspend fun publishSuccess(key: K, load: Load<V>, value: V) = withContext(NonCancellable) {
        mutex.withLock {
            releaseLocked(key, load)
            load.fetched = now()
            sweepLocked()
            makeRoomLocked(key)
            seq++
            entries[key] = Entry(value, load.fetched, seq)
        }
        load.done.complete(value)
    }

    private suspend fun publishFailure(key: K, load: Load<V>, cause: Throwable) = withContext(NonCancellable) {
        // The key is released so a later caller can load again; the previous
        // value, if any, stays in place.
        mutex.withLock { releaseLocked(key, load) }
        load.done.completeExceptionally(cause)
    }

    /**
     * Releases the key BY IDENTITY: the slot is cleared only when the record
     * sitting in it is the one being published.
     *
     * A publisher that removed whatever it found would drop a LATER caller's
     * single-flight guarantee while that caller was still loading — two
     * concurrent loads for one key, and the listing's key is the bare account
     * id, so that is an account-wide duplicate rather than a bucket's.
     *
     * That is not reachable as this cache is written today: a second caller
     * finding the slot occupied becomes a waiter rather than a registrant, and
     * every publish clears the slot BEFORE completing the deferred, so no waiter
     * can wake and re-register ahead of its own publisher. The guard is here
     * anyway because that argument is global and this one is local — it holds by
     * reading four lines rather than by reasoning about every path that could
     * ever publish. Re-introduce an abandonment path and the invariant still
     * stands instead of silently going with it.
     */
    private fun releaseLocked(key: K, load: Load<V>) {
        if (inflight[key] === load) inflight.remove(key)
    }

    /**
     * Drops every entry past its TTL. It runs at each publication — the one
     * moment the cache does work proportional to a miss anyway — so a long-lived
     * client that has seen many buckets keeps a snapshot for at most a TTL past
     * its last use plus the interval to the next load on any key, rather than for
     * its lifetime. Caller holds [mutex].
     */
    private fun sweepLocked() {
        val cutoff = now()
        entries.entries.removeAll { cutoff - it.value.fetched >= ttlMillis }
    }

    /**
     * Evicts the oldest-fetched entries until the one about to be stored for
     * [key] fits under [maxItems] — oldest by fetch time, and by publication
     * order among equals, so the choice is total rather than whatever map
     * iteration happens to visit first. Oldest-first is also the right order: the
     * entry nearest its TTL is the one least worth keeping. Caller holds [mutex].
     */
    private fun makeRoomLocked(key: K) {
        if (maxItems <= 0) return
        if (entries.containsKey(key)) return // an overwrite takes no new room
        while (entries.size >= maxItems) {
            val oldest = entries.entries.minWithOrNull(
                compareBy({ it.value.fetched }, { it.value.seq }),
            ) ?: return
            entries.remove(oldest.key)
        }
    }
}

/** A dock cache key: an account's view of one bucket. */
internal data class CampfireBucketKey(val accountId: String, val bucketId: Long)

/**
 * One consultation of a discovery source: the candidate ids it holds for the
 * bucket, when that snapshot was fetched, and whether it predated the call.
 */
internal class SourceRead(val ids: List<Long>, val fetched: Long, val cached: Boolean)

/**
 * The load failure a Campfire listing past its cap raises;
 * `RecordingsService.summarize` turns it into the typed
 * `campfire_discovery_incomplete`.
 */
internal class CampfireListingOverflow :
    Exception("campfire listing exceeds MAX_CAMPFIRE_LISTING ($MAX_CAMPFIRE_LISTING)")

/**
 * Campfire discovery for chat lines: the two sources, each cached
 * [CAMPFIRE_INDEX_TTL_MILLIS].
 *
 * A `chat.line.created` row carries the line's id and bucket, not its Campfire,
 * and the line read is `/chats/{campfireId}/lines/{lineId}`. Candidates come
 * from two sources, tried in order:
 *
 *  1. The bucket's project dock, whose `chat` tool is the project's Campfire:
 *     one project read per bucket, and the answer for every line posted in a
 *     project. Cached per bucket.
 *  2. The account-wide Campfire listing (BC3 has no per-bucket one), filtered to
 *     the bucket, for buckets that are not projects or whose dock did not hold
 *     the line. Cached per account, so a burst of lines costs one listing.
 *
 * It lives on the client (shared by every [AccountClient] the client hands out);
 * a client is bound to one credential, so entries are never shared across
 * authorization contexts, and every key carries the account id. Expired
 * snapshots are swept at each load and each cache is bounded
 * ([CAMPFIRE_INDEX_MAX_ITEMS], oldest out first), so the index holds at most the
 * buckets and accounts consulted within the last TTL, and never more than the
 * bound, whatever the client has seen.
 */
internal class CampfireIndex(now: () -> Long = ::monotonicMillis) {
    private val docks = TtlCache<CampfireBucketKey, List<Long>>(
        now,
        CAMPFIRE_INDEX_TTL_MILLIS,
        CAMPFIRE_INDEX_MIN_REFRESH_MILLIS,
        CAMPFIRE_INDEX_MAX_ITEMS,
    )
    private val listings = TtlCache<String, Map<Long, List<Long>>>(
        now,
        CAMPFIRE_INDEX_TTL_MILLIS,
        CAMPFIRE_INDEX_MIN_REFRESH_MILLIS,
        CAMPFIRE_INDEX_MAX_ITEMS,
    )

    /**
     * The Campfire ids a bucket's project dock names. A bucket that is not a
     * project (a 404 on the project read) has none; any other failure of the read
     * is raised.
     */
    suspend fun dockCampfires(account: AccountClient, bucketId: Long, refresh: Boolean): SourceRead {
        val key = CampfireBucketKey(account.accountId, bucketId)
        val hit = docks.get(key, refresh) {
            val project = try {
                account.projects.get(bucketId)
            } catch (e: BasecampException.NotFound) {
                null
            }
            project?.dock.orEmpty().filter { it.name == "chat" && it.id != 0L }.map { it.id }
        }
        return SourceRead(hit.value, hit.fetched, hit.cached)
    }

    /**
     * The Campfire ids the cached account-wide listing shows in a bucket, without
     * fetching: null when the listing is not cached or has expired.
     */
    suspend fun cachedListedCampfires(accountId: String, bucketId: Long): SourceRead? {
        val hit = listings.peek(accountId) ?: return null
        return SourceRead(hit.value[bucketId].orEmpty(), hit.fetched, cached = true)
    }

    /**
     * The Campfire ids the account-wide listing shows in a bucket. A listing that
     * overflows [MAX_CAMPFIRE_LISTING] is not cached and raises
     * [CampfireListingOverflow], which the caller reports as incomplete
     * discovery.
     */
    suspend fun listedCampfires(account: AccountClient, bucketId: Long, refresh: Boolean): SourceRead {
        val hit = listings.get(account.accountId, refresh) {
            val listed = account.campfires.list(PaginationOptions(maxItems = MAX_CAMPFIRE_LISTING))
            if (listed.meta.truncated) throw CampfireListingOverflow()
            listed.filter { it.bucket.id != 0L }.groupBy({ it.bucket.id }, { it.id })
        }
        return SourceRead(hit.value[bucketId].orEmpty(), hit.fetched, hit.cached)
    }
}

File: ruby/lib/basecamp/campfire_index.rb
# frozen_string_literal: true

module Basecamp
  # The two Campfire discovery sources a chat line's read needs, each cached.
  #
  # A +chat.line.created+ row carries the line's id and bucket, not its
  # Campfire, and the line read is +/chats/{campfireId}/lines/{lineId}+.
  # Candidates come from two sources, tried in order and each cached {TTL}
  # seconds:
  #
  # 1. The bucket's project dock, whose "chat" tool is the project's Campfire:
  #    one project read per bucket, and the answer for every line posted in a
  #    project. Cached per bucket.
  # 2. The account-wide Campfire listing (BC3 has no per-bucket one), filtered
  #    to the bucket, for buckets that are not projects or whose dock did not
  #    hold the line. Cached per account, so a burst of lines costs one listing.
  #
  # This class holds the caching only; the reads themselves are passed in as
  # blocks by +RecordingsExtensions+, so every request still goes through a
  # generated service method (SPEC section 18 rule 1) and this file touches no
  # wire.
  #
  # It lives on {Client}, shared by every {AccountClient} that client hands out.
  # A Client is bound to one credential, so entries are never shared across
  # authorization contexts, and every key carries the account id anyway. Expired
  # snapshots are swept at each publication and each cache is bounded
  # ({MAX_ITEMS}, oldest fetched out first), so the index holds at most the
  # buckets and accounts consulted within the last TTL, and never more than the
  # bound, whatever the Client has seen.
  class CampfireIndex
    # How long a cached discovery source — a bucket's project dock, the
    # account's Campfire listing — is reused before it is read again.
    TTL = 600.0

    # Bounds the refresh-on-miss: a line found under no candidate re-reads the
    # cached sources, but not more often than this per source, so a run of
    # unresolvable lines cannot turn into a listing per line.
    # +UnresolvedRecordingError#refreshed?+ says whether the floor applied.
    MIN_REFRESH = 30.0

    # Caps the account-wide Campfire listing the fallback source reads. A
    # listing that overflows it is not cached and the call reports
    # {CampfireDiscoveryIncompleteError}: the dock covers every project, so the
    # listing only ever serves the leftover, and an account with more Campfires
    # than this should not pay a full walk per TTL for it.
    MAX_LISTING = 1000

    # Bounds each cache's entry count. The dock cache holds one snapshot per
    # bucket consulted within the TTL: a connector listening across every
    # project an agent can see touches hundreds of buckets, not thousands, and a
    # snapshot is a handful of ids, so this is generous headroom at a few
    # hundred KB. The bound exists so that a process alive for weeks can never
    # grow past it whatever it sees. When the bound is reached the
    # oldest-fetched entries go first, deterministically.
    MAX_ITEMS = 1024

    # Raised by the listing loader when the account-wide listing overflows
    # {MAX_LISTING}. +RecordingsExtensions+ turns it into the typed
    # {CampfireDiscoveryIncompleteError}; it is never cached, so the next call
    # re-reads rather than remembering a verdict it never reached.
    class ListingOverflow < StandardError; end

    # One consultation of a discovery source: the candidate ids it holds for the
    # bucket, when that snapshot was fetched, and whether it predated the call.
    SourceRead = Struct.new(:ids, :fetched, :cached, keyword_init: true) do
      # @return [Boolean] whether the snapshot predated this call
      def cached?
        cached
      end
    end

    # @param clock [#call, nil] monotonic seconds, injectable for tests
    def initialize(clock: nil)
      @docks = TTLCache.new(ttl: TTL, floor: MIN_REFRESH, max_items: MAX_ITEMS, clock: clock)
      @listings = TTLCache.new(ttl: TTL, floor: MIN_REFRESH, max_items: MAX_ITEMS, clock: clock)
    end

    # The Campfire ids a bucket's project dock names.
    #
    # @param account_id [String]
    # @param bucket_id [Integer]
    # @param refresh [Boolean] re-read a cached snapshot older than {MIN_REFRESH}
    # @yieldreturn [Array<Integer>] the dock's Campfire ids
    # @return [SourceRead]
    def dock_campfires(account_id:, bucket_id:, refresh: false, &load)
      hit = @docks.get([ account_id, bucket_id ], refresh: refresh, &load)
      SourceRead.new(ids: Array(hit.value), fetched: hit.fetched, cached: hit.cached)
    end

    # The Campfire ids the CACHED account-wide listing shows in a bucket,
    # without fetching. nil when the listing is not cached or has expired.
    #
    # It exists so a caller can consult what the source already holds before
    # deciding whether to pay for the listing fetch — the expensive, slow
    # request, which the dock (including its refresh) should get to pre-empt.
    #
    # @param account_id [String]
    # @param bucket_id [Integer]
    # @return [SourceRead, nil]
    def cached_listed_campfires(account_id:, bucket_id:)
      hit = @listings.peek(account_id)
      return nil if hit.nil?

      SourceRead.new(ids: Array(hit.value[bucket_id]).dup, fetched: hit.fetched, cached: true)
    end

    # The Campfire ids the account-wide listing shows in a bucket.
    #
    # @param account_id [String]
    # @param bucket_id [Integer]
    # @param refresh [Boolean] re-read a cached snapshot older than {MIN_REFRESH}
    # @yieldreturn [Hash{Integer => Array<Integer>}] campfire ids by bucket id
    # @return [SourceRead]
    # @raise [ListingOverflow] when the loader reports the listing was truncated
    def listed_campfires(account_id:, bucket_id:, refresh: false, &load)
      hit = @listings.get(account_id, refresh: refresh, &load)
      SourceRead.new(ids: Array(hit.value[bucket_id]).dup, fetched: hit.fetched, cached: hit.cached)
    end

    # A per-key cache with single-flight loading: concurrent callers for one key
    # wait on the one load in progress rather than loading again, a failed load
    # leaves the previous value in place (and its error is shared with the
    # waiters, so N callers never re-run one failed load N times), and a refresh
    # is honoured only once the value is older than a floor.
    class TTLCache
      # Raised when a load left without publishing an outcome of its own — the
      # loading thread was killed, or unwound by something that is not a
      # StandardError, +Timeout.timeout+ included (it raises
      # +Timeout::ExitException+ INSIDE the block and only becomes a
      # +Timeout::Error+ at its own frame, so no rescue here ever sees it).
      #
      # It is a {Basecamp::Error} so that +rescue Basecamp::Error+ around a
      # composite catches it like everything else, with +api_error+ and
      # +retryable+ — the load reached no verdict and the key is free, so the
      # next call loads for itself.
      #
      # A waiter does not normally re-run another caller's failed load, but this
      # one it does, once: the classification is unambiguous here in a way Go's
      # two-part +callerDone+ had to work for. A transport failure is a
      # StandardError and is shared as the load's own; only a genuinely
      # non-local unwind of the LOADING thread reaches this, and that says
      # nothing about whether the waiter's own call can succeed. Without the
      # re-run, one caller's +Timeout.timeout+ fails every concurrent waiter on
      # a key the whole account shares.
      class LoaderAbandoned < Error
        def initialize(message = "cache loader did not complete")
          super(code: ErrorCode::API, message: message, retryable: true)
        end
      end

      # What a cache read hands back: the value, when it was fetched, and
      # whether it predated the call (as opposed to being loaded during it, by
      # this caller or by one it waited on). The fetch time is what lets a
      # caller tell a snapshot it already consulted from a newer one, whoever
      # loaded it.
      Hit = Struct.new(:value, :fetched, :cached, keyword_init: true)

      Entry = Struct.new(:value, :fetched, :seq, keyword_init: true)

      # The clock is MONOTONIC, never wall time. A backwards step in wall time —
      # an NTP correction, a VM resume — would make an entry outlive its TTL,
      # decline a refresh that is genuinely due, and take the `refreshed` and
      # stale-candidate signals down with it, none of which surfaces as an
      # error. The injected clock in tests is monotonic in the same sense: it
      # only ever moves forward.
      #
      # @param ttl [Float] seconds a value is reused for
      # @param floor [Float] seconds a refresh must wait before it is honoured
      # @param max_items [Integer] entry bound; 0 or less disables eviction
      # @param clock [#call, nil] monotonic seconds, injectable for tests
      def initialize(ttl:, floor:, max_items:, clock: nil)
        @ttl = ttl
        @floor = floor
        @max_items = max_items
        @clock = clock || -> { Process.clock_gettime(Process::CLOCK_MONOTONIC) }
        @mutex = Mutex.new
        @condition = ConditionVariable.new
        @entries = {}
        @inflight = {}
        @seq = 0
      end

      # Returns the value for key, loading it when absent or older than the TTL
      # — or, with +refresh+ set, older than the floor.
      #
      # @param key [Object]
      # @param refresh [Boolean]
      # @yieldreturn [Object] the loaded value
      # @return [Hit]
      def get(key, refresh: false)
        reloaded = false

        loop do
          owner = false
          pending = nil
          published = false

          # The ensure encloses the REGISTRATION, not just the loader. A loader
          # that leaves by anything the rescue below does not catch — Interrupt,
          # a signal, NoMemoryError, a Thread#kill that raises nothing at all —
          # would otherwise leave the key in flight forever, and since {#await}
          # waits with no timeout, every later caller for that key would park on
          # it permanently rather than erroring. The listing's key is the bare
          # account id, so that is one killed thread wedging chat-line discovery
          # for a whole account for the life of the process. Go publishes from a
          # deferred recover for exactly this reason — and an ensure that began
          # after the key was registered would leave the same hole a few
          # instructions wide.
          begin
            @mutex.synchronize do
              entry = @entries[key]
              if entry
                age = @clock.call - entry.fetched
                if age < @ttl && (!refresh || age < @floor)
                  return Hit.new(value: entry.value, fetched: entry.fetched, cached: true)
                end
              end

              pending = @inflight[key]
              if pending.nil?
                pending = { done: false, error: nil, value: nil, fetched: nil }
                # Ownership is claimed BEFORE the key is registered, so an
                # interrupt between the two leaves the ensure publishing a
                # record that was never registered — which {#publish} ignores,
                # because it only ever evicts its own.
                owner = true
                @inflight[key] = pending
              end
            end

            unless owner
              begin
                return await(pending)
              rescue LoaderAbandoned
                # The load did not fail, it was abandoned — see
                # {LoaderAbandoned}. That says nothing about whether this
                # caller's own call can succeed, so it goes round once and loads
                # for itself. Once only, so a run of abandoned owners cannot
                # become a queue of sequential loads behind one waiter.
                raise if reloaded

                reloaded = true
              end
              next
            end

            value = yield
            publish(key, pending, value, nil)
            published = true
            return Hit.new(value: value, fetched: pending[:fetched], cached: false)
          rescue StandardError => e
            publish(key, pending, nil, e) if owner
            published = true
            raise
          ensure
            publish(key, pending, nil, LoaderAbandoned.new) if owner && !published
          end
        end
      end

      # Returns the cached value for key when one is within the TTL, without
      # loading.
      #
      # @param key [Object]
      # @return [Hit, nil]
      def peek(key)
        @mutex.synchronize do
          entry = @entries[key]
          return nil if entry.nil? || (@clock.call - entry.fetched) >= @ttl

          Hit.new(value: entry.value, fetched: entry.fetched, cached: true)
        end
      end

      private

      # Waits on another caller's load and hands its outcome over.
      #
      # The value is read off the load record, not out of the cache, so a sweep
      # or the entry bound evicting the entry in the meantime cannot take it
      # from this waiter. It is reported as loaded during this call rather than
      # as something that predated it, because it was.
      def await(pending)
        @mutex.synchronize do
          @condition.wait(@mutex) until pending[:done]
        end
        raise pending[:error] if pending[:error]

        Hit.new(value: pending[:value], fetched: pending[:fetched], cached: false)
      end

      # Whether this waiter should load for itself rather than take the outcome
      # it just woke on. Only for a load that was abandoned rather than failed —
      # see {LoaderAbandoned} — and only once per call, so a run of abandoned
      # owners cannot become a queue of sequential loads behind one waiter.
      def reload_after?(pending, reloaded)
        !reloaded && pending[:error].is_a?(LoaderAbandoned)
      end

      # Releases the key and wakes the waiters. A failed load publishes no entry,
      # so the previous value — if any — stays in place and keeps serving until
      # its own TTL runs out.
      def publish(key, pending, value, error)
        @mutex.synchronize do
          # Idempotent, so publication survives being interrupted. An async
          # exception landing between the load returning and this call
          # completing would otherwise let {#get}'s ensure publish a second time
          # and overwrite a good value with an abandonment error. Whichever
          # publication lands first is the outcome; the rest only wake the
          # waiters again.
          #
          # The repeat still broadcasts, and that is the point of doing it here
          # rather than returning bare: an interrupt landing between the flag
          # below and its broadcast would otherwise leave a record marked done
          # with everybody still parked on it.
          if pending[:done]
            @condition.broadcast
            return
          end

          # Only this call's own registration. An abandonment publish for a
          # record that never made it into the map — the interrupt window the
          # ensure covers — would otherwise delete whatever ANOTHER thread has
          # since registered under the key, dropping its single-flight guarantee
          # while it is still loading. The listing's key is the bare account id,
          # so that is process-wide.
          @inflight.delete(key) if @inflight[key].equal?(pending)
          pending[:error] = error
          if error.nil?
            pending[:value] = value
            pending[:fetched] = @clock.call
            sweep_locked
            make_room_locked(key)
            @seq += 1
            @entries[key] = Entry.new(value: value, fetched: pending[:fetched], seq: @seq)
          end
          pending[:done] = true
          @condition.broadcast
        end
      end

      # Drops every entry past its TTL. Runs at each publication — the one
      # moment the cache does work proportional to a miss anyway — so a
      # long-lived Client that has seen many buckets keeps a snapshot for at most
      # a TTL past its last use plus the interval to the next load on any key,
      # rather than for its lifetime.
      def sweep_locked
        now = @clock.call
        @entries.delete_if { |_key, entry| (now - entry.fetched) >= @ttl }
      end

      # Evicts the oldest-fetched entries until the one about to be stored for
      # +key+ fits under the bound — oldest by fetch time, and by publication
      # order among equals, so the choice is total rather than whatever hash
      # order happens to come first. Oldest-first is also the right order: the
      # entry nearest its TTL is the one least worth keeping.
      def make_room_locked(key)
        return if @max_items <= 0
        return if @entries.key?(key) # an overwrite takes no new room

        while @entries.size >= @max_items
          oldest = @entries.min_by { |_k, entry| [ entry.fetched, entry.seq ] }
          break if oldest.nil?

          @entries.delete(oldest.first)
        end
      end
    end
  end
end
